crates/findex/src/implementations/redis/findex.rs

```diff
@@ -0,0 +1,466 @@
+        }
+        pipe.atomic().query_async(&mut self.manager.clone()).await?;
+        Ok(())
+    }
+
+    async fn update_lines(
+        &self,
+        chain_table_uids_to_remove: Uids<UID_LENGTH>,
+        new_encrypted_entry_table_items: EncryptedTable<UID_LENGTH>,
+        new_encrypted_chain_table_items: EncryptedTable<UID_LENGTH>,
+    ) -> Result<(), FindexRedisError> {
+        // Prevent compacting while already compacting
+        // We could add support to make sure we are not upserting
+        let _compact_lock = match self.compact_lock.try_write() {
+            Ok(lock) => lock,
+            Err(_) => return Err(FindexRedisError::Compacting), /* another thread is compacting already so we return an error */
+        };
+
+        trace!(
+            "update_lines chain_table_uids_to_remove: {}, new_encrypted_entry_table_items: {}, \
+             new_encrypted_chain_table_items: {}",
+            chain_table_uids_to_remove.len(),
+            new_encrypted_entry_table_items.len(),
+            new_encrypted_chain_table_items.len()
+        );
+        // collect all the entry table keys to delete
+        let entry_keys_to_delete: Vec<Vec<u8>> = self
+            .manager
+            .clone()
+            .keys(key(FindexTable::Entry, b"*"))
+            .await?;
+
+        // add new entry table entries
+        // keep track of the keys we added so we don't delete them later on
+        let mut new_keys: Vec<Vec<u8>> = Vec::with_capacity(new_encrypted_chain_table_items.len());
+        let mut pipeline = pipe();
+        for item in new_encrypted_entry_table_items {
+            let key = key(FindexTable::Entry, &item.0);
+            pipeline.set(key.clone(), item.1);
+            // do not delete that key later on
+            new_keys.push(key);
+        }
+        pipeline
+            .atomic()
+            .query_async(&mut self.manager.clone())
+            .await?;
+
+        // delete the chain table
+        let mut pipeline = pipe();
+        for item in chain_table_uids_to_remove {
+            pipeline.del(key(FindexTable::Chain, &item));
+        }
+        pipeline
+            .atomic()
+            .query_async(&mut self.manager.clone())
+            .await?;
+
+        // add new chain table entries
+        let mut pipeline = pipe();
+        for item in new_encrypted_chain_table_items {
+            pipeline.set(key(FindexTable::Chain, &item.0), item.1);
+        }
+        pipeline
+            .atomic()
+            .query_async(&mut self.manager.clone())
+            .await?;
+
+        // now delete the old entries
+        let mut pipeline = pipe();
+        for entry_key in entry_keys_to_delete {
+            // do not delete the new keys
+            if new_keys.contains(&entry_key) {
+                continue;
+            }
+            pipeline.del(entry_key);
+        }
+        pipeline
+            .atomic()
+            .query_async(&mut self.manager.clone())
+            .await?;
```

Comment:
The order is still wrong. You should:

1. save old entry UIDs
2. add new chains
3. add new entries
4. publish the new label
5. delete old entries
6. delete old chains

That way, you always maintain a working system in case of failure (cf the [documentation](https://github.com/Cosmian/findex/blob/fce3735f76e7fcb4c2d505205f74ac55b2052407/src/callbacks.rs#L122) of the option 2 of the upsert function in Findex).

Reply:
Done in 7867bf8